test/intergration_test.go

```diff
@@ -0,0 +1,225 @@
+	web.Config.Base = "/dbs"
+	web.Config.DBFile = "../dbfile"
+	web.Config.LexiconFile = lexiconFile
+	web.Config.ServerCrt = ""
+	web.Config.ServerKey = ""
+	web.Config.ServerType = serverType
+	web.Config.LogFile = "/tmp/dbs2go-test.log"
+	web.Config.Verbose = 0
+
+	initTestLimiter(t, "100-S")
+}
+
+// injects dbs records
+func injectDBRecord(t *testing.T, rec dbs.DBRecord, method string, url string, endpoint string, handler func(http.ResponseWriter, *http.Request)) []dbs.Record {
+	data, err := json.Marshal(rec)
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+	reader := bytes.NewReader(data)
+	req := newreq(t, method, url, endpoint, reader, nil)
+	req.Header.Add("Accept", "application/json")
+	req.Header.Set("Content-Type", "application/json")
+	r, err := http.DefaultClient.Do(req)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer r.Body.Close()
+
+	if r.StatusCode != http.StatusOK {
+		t.Fatal(err)
+	}
+
+	rr, err := respRecorder("GET", endpoint, nil, handler)
+	if err != nil {
+		t.Error(err)
+	}
+
+	var records []dbs.Record
+	data = rr.Body.Bytes()
+	err = json.Unmarshal(data, &records)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return records
+}
+
+// fetches data from url and endpoint
+func getData(t *testing.T, url string, endpoint string, params url.Values) ([]dbs.Record, int) {
+	r, err := http.DefaultClient.Do(newreq(t, "GET", url, endpoint, nil, params))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer r.Body.Close()
+
+	if r.StatusCode != http.StatusOK {
+		t.Fatalf("Bad status code: %v", r.StatusCode)
+	}
+
+	var d []dbs.Record
+	err = json.NewDecoder(r.Body).Decode(&d)
+	if err != nil {
+		t.Fatalf("Failed to decode body, %v", err)
+	}
+
+	return d, r.StatusCode
+}
+
+// creates an http request for testing
+func newreq(t *testing.T, method string, hostname string, endpoint string, body io.Reader, params url.Values) *http.Request {
+	url2, err := url.Parse(hostname)
+	if err != nil {
+		t.Fatal(err)
+	}
+	url2.Path = endpoint
+	url2.RawQuery = params.Encode()
+
+	r, err := http.NewRequest(method, url2.String(), body)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return r
+}
+
+// TestDBSIntegration Tests both DBSReader and DBSWriter Endpoints
+func TestDBSIntegration(t *testing.T) {
+	db := initDB(false)
+	defer db.Close()
+
+	lexiconFileWriter := os.Getenv("DBS_WRITER_LEXICON_FILE")
+	if lexiconFileWriter == "" {
+		t.Fatal("no DBS_WRITER_LEXICON_FILE env variable, please define")
+	}
+
+	lexiconFileReader := os.Getenv("DBS_READER_LEXICON_FILE")
+	if lexiconFileReader == "" {
+		t.Fatal("no DBS_READER_LEXICON_FILE env variable, please define")
+	}
+
+	configTestServer(t, "DBSWriter", lexiconFileWriter)
+	// start DBSWriter server
+	tsW := httptest.NewServer(web.Handlers())
+	defer tsW.Close()
+
+	configTestServer(t, "DBSReader", lexiconFileReader)
+	// start DBSReader server
+	tsR := httptest.NewServer(web.Handlers())
+	defer tsR.Close()
+
+	t.Run("Test datatiers", func(t *testing.T) {
```

Comment:
This part will repeats over and over and must be generalized. How about the following changes:

```
func verifyResponse(t *testing.T, received []dbs.Record, emap map[string]interface{}) {
        for _, r := range received {
                for k, v := range emap {
                        val := r[k]
                        if val != v {
                                t.Fatalf("Incorrect %s: Expected %v, Received %v", k, v, val)
                        }
                }
        }
}

func runTestWorkflow(t *testing.T, tsR *httptest.Server, tsW *httptest.Server, api string, dbrec dbs.DBRecord
, params url.Values) {
        data, err := json.Marshal(dbrec)
        if err != nil {
                t.Fatal(err.Error())
        }
        var emap map[string]interface{}
        err = json.Unmarshal(data, &emap)
        if err != nil {
                t.Fatal(err.Error())
        }
        t.Run("Test empty GET", func(t *testing.T) {
                d, _ := getData(t, tsR.URL, api, nil)
                if len(d) != 0 {
                        t.Fatal("Data exists")
                }
        })

        t.Run("Test POST", func(t *testing.T) {
                records := injectDBRecord(t, dbrec, "POST", tsW.URL, api, web.DatatiersHandler)
                verifyResponse(t, records, emap)
        })

        t.Run("Test GET after POST", func(t *testing.T) {
                d, _ := getData(t, tsR.URL, api, nil)
                verifyResponse(t, d, emap)
        })
        t.Run("Test GET with parameters", func(t *testing.T) {
                getData(t, tsR.URL, api, params)
        })
}
```
and then you can use it like this:
```
        dt := &dbs.DataTiers{
                DATA_TIER_NAME: "GEN-SIM-RAW",
                CREATE_BY:      "tester",
        }
        api := "/dbs/datatiers"
        params := url.Values{}
        params.Add("data_tier_name", "GEN-SIM-RAW")
        runTestWorkflow(t, tsR, tsW, api, dt, params)

        pg := &dbs.PhysicsGroups{
                PHYSICS_GROUP_NAME: "Tracker",
        }
        api = "dns/physicsgroup"
        params = url.Values{}
        params.Add("physics_group_name", "Tracker")
        runTestWorkflow(t, tsR, tsW, api, pg, params)
```
```